build/Shared/EqualityUtility.cs

```diff
@@ -112,14 +166,93 @@ internal static class EqualityUtility
                 return true;
             }
 
+            if (sequenceComparer == null)
+            {
+                sequenceComparer = EqualityComparer<TSource>.Default;
+            }
+
             if (self.Count == 1)
             {
-                return (sequenceComparer ?? EqualityComparer<TSource>.Default).Equals(self[0], other[0]);
+                return sequenceComparer.Equals(self[0], other[0]);
+            }
+
+            TSource[] selfCopy = new TSource[self.Count];
+            self.CopyTo(selfCopy, 0);
+            Array.Sort(selfCopy, orderComparer);
+
+            TSource[] otherCopy = new TSource[other.Count];
+            other.CopyTo(otherCopy, 0);
+            Array.Sort(otherCopy, orderComparer);
+
+            for (int i = 0; i < selfCopy.Length; ++i)
+            {
+                if (!sequenceComparer.Equals(selfCopy[i], otherCopy[i]))
+                {
+                    return false;
+                }
             }
 
-            return self
-                .OrderBy(keySelector, orderComparer)
-                .SequenceEqual(other.OrderBy(keySelector, orderComparer), sequenceComparer);
+            return true;
+        }
+
+        /// <summary>
+        /// Compares two lists for equality, ordered according to the specified key and optional comparer. Handles null values gracefully.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the list</typeparam>
+        /// <param name="self">This list</param>
+        /// <param name="other">The other list</param>
+        /// <param name="keySelector">The function to extract the key from each item in the list</param>
+        /// <param name="orderComparer">An optional comparer for comparing keys</param>
+        /// <param name="sequenceComparer">An optional comparer for sequences</param>
+        internal static bool OrderedEquals<TSource>(this IReadOnlyList<TSource>? self, IReadOnlyList<TSource>? other, Comparison<TSource> orderComparer, IEqualityComparer<TSource>? sequenceComparer = null)
+        {
+            Debug.Assert(sequenceComparer != null || typeof(TSource) != typeof(string), "Argument " + "sequenceComparer" + " must be provided if " + "TSource" + " is a string.");
+
+            if (TryIdentityEquals(self, other, out bool identityEquals))
+            {
+                return identityEquals;
+            }
+
+            if (self.Count != other.Count)
+            {
+                return false;
+            }
+
+            if (self.Count == 0)
+            {
+                return true;
+            }
+
+            if (sequenceComparer == null)
+            {
+                sequenceComparer = EqualityComparer<TSource>.Default;
+            }
+
+            if (self.Count == 1)
+            {
+                return sequenceComparer.Equals(self[0], other[0]);
+            }
+
+            TSource[] selfCopy = new TSource[self.Count];
```

Comment:
The array comparisons are equivalent in the Enumerable, IList, IReadOnlyList implementations.
Can we create a method to avoid accidental issues?